WIP: Multi region formation

rjocoleman opened the pull request:
This is a WIP.

It's matched with https://github.com/convox/cli/pull/47

It:

* Decouples Lambda Cloudformation bridge from the main Cloudformation Stack (so they can be created in different regions).
* Adds an Input Parameter in the `kernel.json` cloudformation stack for the Lambda Cloudformation Bridge ARN.
* Support SNS as a notification source for the Lambda bridge function

TODO:

- [ ] https://github.com/convox/kernel/blob/master/cmd/formation/handler/aws.go#L54
- [ ] Support regions with less than three AZs

rjocoleman commented:
@ddollar 

Supporting arbitrary numbers of AZs is proving to be a bit of a pain.

I assume that `Custom::EC2AvailabilityZones` exists because creating subnets is unreliable and inconsistent.

Conditions should be the way forward but they don't support the output of Resources like `Custom::EC2AvailabilityZones`.

How about this:
`Custom::EC2Subnet`
Return values: Array of created subnet ids and comma separated list.

`Custom::SubnetRouteTableAssociation`
Return value: Array of resource names 

I'd also add an array return value to `Custom::EC2AvailabilityZones` that can be used inplace of the hard coded arrays.

The goal would be support of subnets in 2-many AZs. 

Here are the problematic lines I'm looking to replace with the above output values:
https://github.com/convox/kernel/blob/master/dist/kernel.json#L301
https://github.com/convox/kernel/blob/master/dist/kernel.json#L339
https://github.com/convox/kernel/blob/master/dist/kernel.json#L456-L465
https://github.com/convox/kernel/blob/master/dist/kernel.json#L598-L602
https://github.com/convox/kernel/blob/master/dist/kernel.json#L803

rjocoleman commented:
Update: I've got the subnets and route associations creating, updating and deleting dynamically via the Lambda bridge as they should.

I'm having trouble dealing with the variable length input coming back for usage when creating the ELB and the ASG. 

I've got a ticket in with AWS support to see if they can shed some light, I've got some other half formed ideas here too so I'll finish it off in the morning.

rjocoleman commented:
this is done, workable solution in f7b0b8979cde16e299875c70900a46a560342e54

Here's what's up:

* I can't find a way to pass back in an JSON array from the custom resource `[ "us-east-1a",  "us-east-1b",  "us-east-1c" ]`
* Cloudformation doesn't like null or empty values.
* Cloudformation's conditional features only work with Conditions
* Conditions can't be used with resources.

So my workaround is:
1. Specify _at least_ the number of results we want, in this case:

```json
        "Subnets": [
          { "Fn::GetAtt": [ "Subnets", "SubnetId0" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId1" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId2" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId3" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId4" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId5" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId6" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId7" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId8" ] },
          { "Fn::GetAtt": [ "Subnets", "SubnetId9" ] }
        ]
```

2. Have `formation` return the same number of subnets. Cloudformation deals with duplication but not empty, null or not present values so I set all of these results to the first.
3. Return the comma separated lists rather than generating them in CFN template.

I've also:

* Using a more flexible value type on the Response map `map[string]interface{}` over `map[string]string`
* Switched `Custom::EC2AvailabilityZones` for a `Custom::EC2Subnets` that creates the subnets based on the AZs that can have subnets created in them.
* Added `Custom::EC2SubnetRouteTableAssociation` to allocate those subnets dynamically.
* Both of these new custom resources also handle deletion but nothing on update (I think this is correct behaviour)
* Removed `AvailabilityZones` from `AWS::AutoScaling::AutoScalingGroup` (it's not required https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-as-group.html#cfn-as-group-vpczoneidentifier)
* Dealt with a race condition deleting subnets that didn't appear using CFN directly.

This paired with the CLI PR this works for me deploying (and undeploying) clusters into non-`us-east-1` regions.

ddollar commented:
This is looking really good! I'm going to hold off for 1-2 more days until we get an answer back from our AWS contacts about wider Lambda availability.

nzoschke commented:
Word is that there is no specific timeline for Lambda in other regions. So we should proceed with the SNS based workaround to add support for other regions.

rjocoleman commented:
Ok, in that case its important to note that this is all held up by: https://github.com/convox/app/pull/11

rjocoleman commented:
@ddollar @nzoschke 

Now that multi-region for Lambda regions has hit I'm keen to refactor my PRs to simplify and support non-lambda regions too. (I'm in `ap-southeast-2`!)

There are two major things to support these:

1. making custom resources via CFN work in regions we can't access lambda
2. supporting < 3 subnets (some regions have less than 3 https://aws.amazon.com/about-aws/global-infrastructure/) (this is already in my PRs, just needs the conflicts resolved)

How do you feel about this workflow?

Install: 

* Create a second bootstrap stack (*first*).
* This stack is conditionally created in the deploy region [if lambda is supported](https://github.com/convox/cli/blob/master/convox/install.go#L83) else in `us-east-1`.
* This stack makes a bootstrap lambda function in with the same formation.zip.
* It has a custom resource that returns `StackCustomTopic`.
* `StackCustomTopic` is an ARN to the lambda function, or it's an SNS topic with the bootstrap lambda function subscribed.
* All the permissions setting and subscriptions happen in this custom resource not the CLI.
* `StackCustomTopic` is passed to the main CFN stack and install happens in the selected deployment region.

Applications:

* `CustomTopic` as currently implemented is removed from the app template.
* Replaced with a custom resource that calls `StackCustomTopic` and returns `AppCustomTopic`.
* `AppCustomTopic` creates a lambda function for the app and conditionally a SNS topic and returns the appropriate ARN.

Installs will end up with 2 cloudformation stacks.
Apps get one each.

nzoschke commented:
What is the simplest thing possible for now? Lambda will be available someday.

Can we go back to when installs had an SNS / lambda handler for all Apps?

Can we use deletion policy retain to keep things around through a kernel delete?

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html

nzoschke commented:
To summarize this PR and some chats:

The installer will make two stacks: a Lambda and SNS stack in a supported region (e.g. us-east-1), and a VPC in the desired region (ap-southeast-2)

The Lambda / SNS topic will be set as a parameter for the VPC stack to use as a custom resource topic.

The Lambda / SNS stack will not be deleted on an `convox uninstall`, so that app stacks can still use the custom resource handler for deletes through cloudformation.

When Lambda is available in every region, we can re-evaluate a strategy that uses a Lambda function in every stacks.

rjocoleman commented:
@nzoschke 

I've done the following:
* Rebased https://github.com/convox/app/pull/11
* Rebased https://github.com/convox/kernel/pull/95 and added the S3 correct bucket.
* Replaced the CLI PR with https://github.com/convox/cli/pull/82

## Details

### `$ convox install`:
  1. creates bootstrap stack containing lambda function called `bootstrap`.
  2. creates SNS topic called `BootstrapTopic`
  3. subscribes `bootstrap` to `BootstrapTopic`.
  4. Creates permissions for `BootstrapTopic` to invoke `bootstrap`.

If it's a rack install region is lambda supported then these all exist in that region.
If the rack install region is not lambda supported then `bootstrap` is created in `us-east-1` and `BootstrapTopic` is created in the rack install region.

This is done outside of CFN and in the cli because a) `AddPermission` is not supported in CFN and b) the subscription needs to exist before the CFN is run.

There is no teardown handling for `bootstrap` and `BootstrapTopic`. I have not changed the name in the later kernel CFN template from `CustomTopic` but they're the same thing.

### kernel

* `kernel.json` has had the lambda extracted from it and put into a new `bootstrap.json`.
* `bootstrap.json` has been added to the `Makefile`.
* `formation` has had processing for SNS events added.
* CustomResources have been created to create subnets and relations to cope with regions that have less than 3 AZs. `kernel.json` has been updated to fit this.

### app

* Use the single rack `BootstrapTopic`/CustomTopic` rather than a discrete one per app.


I haven't done any docs but it should be as simple as `$ convox install --region ap-southeast-2` where `ap-southeast-2` is a non-lambda region.

nzoschke commented:
Thanks @rjocoleman. I'll start testing this now.

nzoschke commented on dist/bootstrap.json:
```diff
@@ -0,0 +1,59 @@
+      }
+    },
+    "Bootstrap": {
+      "Type": "AWS::Lambda::Function",
```
I think we want a DeletionPolicy: Retain to not totally orphan app stacks when we remove the kernel.

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html

nzoschke commented:
In trying to install in us-west-1 I got the following error:

```
Installing Convox...
Lambda is not supported in us-west-1, creating bootstrap in us-east-1:
Created Lambda Function: convox-bootstrap-Bootstrap-1W88F05LOM8RS
Created CloudFormation Stack: convox-bootstrap
Continuing normal install...
Failed AWS::IAM::Role: Resource creation cancelled
Failed AWS::DynamoDB::Table: Resource creation cancelled
Failed AWS::S3::Bucket: Resource creation cancelled
Failed AWS::EC2::VPC: Resource creation cancelled
Failed AWS::DynamoDB::Table: Resource creation cancelled
Failed AWS::IAM::Role: Resource creation cancelled
Failed AWS::S3::Bucket: Resource creation cancelled
Failed Custom::KMSKey: Resource creation cancelled
Failed AWS::Kinesis::Stream: Resource creation cancelled
Failed AWS::EC2::InternetGateway: Resource creation cancelled
Created IAM User: convox-RegistryUser-CQGBQJ9HC7IP
Created IAM User: convox-LogsUser-RV72RP6Z8V7Q
Created IAM User: convox-KernelUser-1X5ELNJ2TQ1KG
Failed AWS::ECS::Cluster: Resource is not supported in this region
Deleted IAM User: convox-RegistryUser-CQGBQJ9HC7IP
Deleted IAM User: convox-LogsUser-RV72RP6Z8V7Q
Deleted IAM User: convox-KernelUser-1X5ELNJ2TQ1KG
ERROR: stack deletion failed
```

I am reaching out to AWS to see why CloudFormation in us-west-1 doesn't support ECS.
